crates/oxc_regular_expression/src/display.rs

```diff
@@ -0,0 +1,368 @@
+                write!(f, "{prefix}i")?;
+            }
+            if flags.sticky {
+                write!(f, "{prefix}y")?;
+            }
+            if flags.multiline {
+                write!(f, "{prefix}m")?;
+            }
+            Ok(())
+        }
+
+        write!(f, "(?")?;
+        if let Some(enabling) = &self.enabling_modifiers {
+            write_flags(f, '\0', enabling)?;
+        }
+        if let Some(disabling) = &self.disabling_modifiers {
+            write_flags(f, '-', disabling)?;
+        }
+        write!(f, ":{})", self.body)
+    }
+}
+
+impl Display for IndexedReference {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "\\{}", self.index)
+    }
+}
+
+impl<'a> Display for NamedReference<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "\\k{}", self.name)
+    }
+}
+
+fn write_join<S, I, E>(f: &mut fmt::Formatter<'_>, sep: S, items: I) -> fmt::Result
+where
+    S: AsRef<str>,
+    E: Display,
+    I: IntoIterator<Item = E>,
+{
+    let sep = sep.as_ref();
+    let mut iter = items.into_iter();
+
+    if let Some(first) = iter.next() {
+        write!(f, "{first}")?;
+    }
+
+    for it in iter {
+        write!(f, "{sep}{it}")?;
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use oxc_allocator::Allocator;
+
+    static CASES: &[(&str, /* expected display */ Option<&str>)] = &[
+        ("/ab/", None),
+        ("/abc/i", None),
+        ("/emo👈🏻ji/u", None),
+        ("/ab|c/i", None),
+        ("/a|b+|c/i", None),
+        ("/(?=a)|(?<=b)|(?!c)|(?<!d)/i", None),
+        (r"/(cg)(?<n>cg)(?:g)/", None),
+        (r"/^(?=ab)\b(?!cd)(?<=ef)\B(?<!gh)$/", None),
+        (r"/^(?<!ab)$/", None),
+        (r"/[abc]/", None),
+        (r"/[a&&b]/v", None),
+        (r"/[a--b]/v", None),
+        (r"/[^a--b--c]/v", None),
+        (r"/[a[b[c[d[e[f[g[h[i[j[k[l]]]]]]]]]]]]/v", None),
+        (r"/[\q{abc|d|e|}]/v", None),
+        (r"/\p{Basic_Emoji}/v", None),
+        // we lose the flags ordering
+        ("/abcd/igv", Some("/abcd/giv")),
+        (r"/\d/ug", Some(r"/\d/gu")),
+        // we always display "syntax characters" as their escaped form.
+        (r"/c]/", Some(r"/c\]/")),
+        (r"/[|\]]/", Some(r"/[\|\]]/")),
+        ("/a{0}|b{1,2}|c{3,}/i", None),
+        // NOTE: surrogated characters can't be displayed without access to the next character so we
+        // can't print them correctly via `Display` trait. Instead we print their unicode code point
+        (r"/Em🥹j/", Some(r"/Em\uD83E\uDD79j/")),
+        // TODO: currently we do **NOT** respect the `Character::kind` field so everything is
+        // lowered to their final form.
+        (r"/\n\cM\0\x41\./", Some("/\n\r\0A\\./")),
+        (r"/\n\cM\0\x41\u1234\./u", Some("/\n\r\0Aሴ\\./u")),
+        (r"/\n\cM\0\x41\u{1f600}\./u", Some("/\n\r\0A😀\\./u")),
+        // TODO: `\b` is parsed as `\u{8}` which is wrong.
+        // (r"/[\bb]/", Some(r"/[\bb]/")),
```

Comment:
@leaysgur  Is this a bug or am I missing something and they represent the same thing?

Reply:
RegExp `[\bb]` is:

- CharacterClass: `[`
  - `\b`: ClassEscape > b = [U+0008 (BACKSPACE)](https://tc39.es/ecma262/2024/multipage/text-processing.html#sec-patterns-static-semantics-character-value)
  - `b`: SourceCharacter
- `]`

I think you need to check `Character.kind == SingleEscape`.

Reply:
Thank you Sensei!

I wasn't sure about this, I think I have to consider `Character::kind` anyway to prevent our prettier and minifier from exploding the regex size or we have to keep the raw string for a while.

I mark this PR as a draft so I can go in and apply the suggested changes and take care of the `Character` displaying.
Surrogated Unicodes bloat the printer regex by quite a few characters so I should also fix this issue by peeking the next character.

Reply:
Yes, checking `Character::kind` is crucial. Otherwise you can not distinguish `\a` and `a`. 😅 

BTW, adding `raw` property to `ast::Character` like other regex parsers may help you in this scenario?

Reply:
Yes, You are right.
I'll add a few more test cases to better test different character kinds.

> BTW, adding raw property to ast::Character like other regex parsers may help you in this scenario?

It definitely does but I think we can do without it; In that case, it would be better as we can keep the memory footprint small. If we ever want to add a raw string it would be best to keep it for the whole pattern, We can correlate characters to this string using their spans. That way we can print the regex literals without any change.

We might want to keep the raw value for the formatter but the minifier should process regex and output the optimal version of it and wouldn't need the raw value, A custom printer that focuses on output size instead of being close to the actual input should do well for it.

Let me try to achieve this without adding any raw strings, If I fail I'll come back to you so we can figure out if and where we should add this raw value.